File: Clustering/DBSCAN.jl
using Clustering
using JLD 

# Create a d*n adjacency matrix to.
function create_adjacency_matrix(doppler_data::AbstractMatrix, range_step::Number, velocity_step::Number)
    total_entries = shape(doppler_data, 1) * shape(doppler_data, 1)
    # The 4 dimensions are: distance, velocity, real, imagenary.
    adjacency_matrix = Matrix{Float32}(undef, 4, total_entries)

end

# Plot the DBSCAN result.
function plot(result::DbscanResult)
     
end

# Meta data.
folder = "Test"
file_number = "012"

# Load the data.
file = "Data/" * folder * "/B210_SAMPLES_" * folder * "_" * file_number * ".jld"
file_data = load(file)
doppler_fft_matrix = file_data["Doppler FFT Matrix"]
range_step = step(file_data["Range"])
velocity_step = step(file_data["Velocity"])

# DBSCAN.
adjacency_matrix = create_adjacency_matrix(doppler_fft_matrix, range_step, velocity_step)
result = dbscan(distance_matrix, 0.5, 4)
display(result)

File: Classification/DBSCAN.py
# Imports.
import h5py
from sklearn.cluster import DBSCAN
import numpy as np
 
# Meta data.
folder = "Test"
file_number = "012"

# Load the data.
file = "Data/" + folder + "/B210_SAMPLES_" + folder + "_" + file_number + ".jld"
file = h5py.File(file, "r")
file_data =  file["Doppler FFT Matrix"]
raw_data = np.matrix(np.array(file_data))

# Convert data to complex type.
complex_data = np.zeros(shape=raw_data.shape, dtype=complex)
for r, row in enumerate(raw_data):
    for c, value in enumerate(row.T):
        complex_data[r][c] = complex(value['re_'], value['im_'])

# Implement DBSCAN.
result = DBSCAN().fit(np.absolute(complex_data))
print(result.labels_)
